### PR TITLE
Deprecate `TypeUtils::getDirectClassNames()`

### PR DESCRIPTION
**Pull request by herndlm**
Just a simple migration over to Type::, as requested in https://github.com/phpstan/phpstan-src/pull/1922#issuecomment-1291875264

Are you fine with that name? E.g. we also have `Type::getReferencedClasses()` which is somehow related, right?

I'd try to get rid of more `instanceof TypeWithClassName` in a follow-up if that's fine

### CODE REVIEW

**Comment by herndlm** [comment]
Looks likes a deprecated method rule in the deprecation rules extension is causing a deprecation now that needs fixing then 🙃

**Comment by herndlm** [comment]
The new failing tests here are from the deprecation extension which needs fixing afterwards. But it would be ready

File: src/Type/UnionTypeHelper.php [hunk 1]
```diff
@@ -32,6 +34,24 @@ public static function getReferencedClasses(array $types): array
...
+	/**
+	 * @param Type[] $types
+	 * @return list<string>
+	 */
+	public static function getObjectClassNames(array $types): array
+	{
+		return array_values(
+			array_unique(
+				array_merge(
+					...array_map(
```
Comment (ondrejmirtes) [inline src/Type/UnionTypeHelper.php]: I really don't like this notation, a good old foreach would be much better. I have no idea what's the result of array_merge with `...array_map` inside :)

As a bonus, with foreach, I can easily put some debugging statements between the calls.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Yeah, the name is OK :) Additionally, we could get rid of `instanceof TypeWithClassName` in this PR. Thank you.

**Comment by herndlm** [comment]
getting rid of all `instanceof TypeWithClassName` here can get ugly really fast. currently there are 40 of those and with many I'm not sure if we should really use the new method, introduce something else, or just leave it as it is.

e.g. some examples

calls where we explicitly want to know the class for whatever reasons
```php
$classType = TypeTraverser::map($classType, static function (Type $type, callable $traverse) use (&$uncertainty): Type {
	if ($type instanceof UnionType || $type instanceof IntersectionType) {
		return $traverse($type);
	}
	if ($type instanceof TypeWithClassName) {
		$uncertainty = true;
		return $type;
	}
...
```
```php
if (!$returnType instanceof TypeWithClassName) {
	return new MixedType();
}
```

calls where we want to know explicitly the **one** className
```php
$thisType instanceof TypeWithClassName ? $thisType->getClassName() : null,
```
```php
$isThrowable = $originalCatchType instanceof TypeWithClassName && strtolower($originalCatchType->getClassName()) === 'throwable';
```
```php
 elseif (
	$argValueType instanceof GenericClassStringType
	&& $argValueType->getGenericType() instanceof TypeWithClassName
) {
	$scopeClass = $argValueType->getGenericType()->getClassName();
	$thisType = $argValueType->getGenericType();
}
```

I'm not entirely sure how to get rid of them yet, e.g. in some cases I'd need to check that `getObjectClassNames()` is not empty, in some I'd need to check that one entry is returned, in some I can maybe traverse them and adapt the surrounding logic (but many places seem to be using `GenericTypeVariableResolver::getType()` which expects a `TypeWithClassName`) and in some maybe I want to check that only one entry is returned and use that. any ideas? or should we do this in more PRs maybe? I want to avoiding messing this one up too much :)

**Comment by ondrejmirtes** [comment]
It shouldn't be hard to come up with examples where those `instanceof TypeWithClassName` are wrong - for example having an intersection with another type (but the condition where we're asking we'd still want it to be true).

Also - UnionType::getObjectClassNames() doesn't match how it should behave. If we for example have Foo|null, the returned array should be empty. Look at getArrays() for an example how to do it.

**Comment by ondrejmirtes** [comment]
You can now just call `pickTypes` in UnionType :) https://github.com/phpstan/phpstan-src/commit/480626ecb52d2e98cc28cee8a18dfb86112b7f8f

**Comment by herndlm** [comment]
> You can now just call `pickTypes` in UnionType :) [480626e](https://github.com/phpstan/phpstan-src/commit/480626ecb52d2e98cc28cee8a18dfb86112b7f8f)

that's great! I just realised that you type-hinted `T of Type` but here T would be `string` :/
can I just remove it or should I make it a `T of Type|string` or do you have some other idea?

**Comment by ondrejmirtes** [comment]
@herndlm Nothing might break if we remove the `of Type` part?

**Comment by herndlm** [comment]
> @herndlm Nothing might break if we remove the `of Type` part?

no, but it starts to get weird naming-wise. `pickTypes()`, `$getTypes`, `$innerTypes` and `$types` are all kind of wrong then

**Comment by ondrejmirtes** [comment]
Yeah, you can rename that to something else :) Also maybe `pickFromTypes` instead of `pickTypes`.

File: src/Type/Php/IsAFunctionTypeSpecifyingHelper.php [hunk 2]
```diff
@@ -25,16 +25,22 @@ public function determineType(
...
 		return TypeTraverser::map(
 			$classType,
-			static function (Type $type, callable $traverse) use ($objectOrClassTypeClassName, $allowString, $allowSameClass): Type {
+			static function (Type $type, callable $traverse) use ($objectOrClassTypeClassNames, $allowString, $allowSameClass): Type {
 				if ($type instanceof UnionType || $type instanceof IntersectionType) {
 					return $traverse($type);
 				}
 				if ($type instanceof ConstantStringType) {
-					if (!$allowSameClass && $type->getValue() === $objectOrClassTypeClassName) {
+					if (!$allowSameClass && $objectOrClassTypeClassNames === [$type->getValue()]) {
```
Comment (staabm) [inline src/Type/Php/IsAFunctionTypeSpecifyingHelper.php:44]: `array_unique` preserves keys. I think therefore this comparison might not work, even when equal named classes are compared because keys might be different?

https://3v4l.org/raB2N

Comment (herndlm) [reply src/Type/Php/IsAFunctionTypeSpecifyingHelper.php:44]: true, thx. actually this is one of the weird cases where it's only relevant if it's one value. I don't quite like my current solution and might come back to this again..

**Comment by ondrejmirtes** [comment]
Nice, I'd merge this if it wasn't a draft 👍

**Comment by herndlm** [comment]
Soon 😊

File: src/Type/BenevolentUnionType.php [hunk 3]
```diff
@@ -43,11 +43,11 @@ protected function unionTypes(callable $getType): Type
 		return TypeUtils::toBenevolentUnion(TypeCombinator::union(...$resultTypes));
 	}
 
-	protected function pickTypes(callable $getTypes): array
+	protected function pickFromTypes(callable $getValues): array
 	{
 		$types = [];
 		foreach ($this->getTypes() as $type) {
-			$innerTypes = $getTypes($type);
+			$innerTypes = $getValues($type);
```
Comment (herndlm) [inline src/Type/BenevolentUnionType.php]: Rename those vars too

File: src/Type/StringAlwaysAcceptingObjectWithToStringType.php [hunk 4]
```diff
@@ -10,19 +10,25 @@ class StringAlwaysAcceptingObjectWithToStringType extends StringType
...
+		$thatClassNames = $type->getObjectClassNames();
+		if ($thatClassNames === []) {
+			return parent::accepts($type, $strictTypes);
 		}
 
-		return parent::accepts($type, $strictTypes);
+		return TrinaryLogic::createNo()->lazyOr(
+			$thatClassNames,
+			static function (string $thatClassName) {
+				$reflectionProvider = ReflectionProviderStaticAccessor::getInstance();
```
Comment (herndlm) [inline src/Type/StringAlwaysAcceptingObjectWithToStringType.php:21]: Move this out of callable

File: src/Type/StringType.php [hunk 5]
```diff
@@ -98,19 +98,25 @@ public function accepts(Type $type, bool $strictTypes): TrinaryLogic
...
+		$thatClassNames = $type->getObjectClassNames();
+		if ($thatClassNames === [] || $strictTypes) {
+			return TrinaryLogic::createNo();
 		}
 
-		return TrinaryLogic::createNo();
+		return TrinaryLogic::createNo()->lazyOr(
+			$thatClassNames,
+			static function (string $thatClassName) {
+				$reflectionProvider = ReflectionProviderStaticAccessor::getInstance();
```
Comment (herndlm) [inline src/Type/StringType.php]: Move this out of callable

File: src/Type/UnionType.php [hunk 6]
```diff
@@ -895,22 +902,34 @@ protected function unionTypes(callable $getType): Type
 	 * @template T of Type
 	 * @param callable(Type $type): list<T> $getTypes
 	 * @return list<T>
+	 *
+	 * @deprecated Use pickFromTypes() instead.
```
Comment (herndlm) [inline src/Type/UnionType.php:916]: weird, BC check was complaining when I removed the whole protected method initially. wonder why. UPDATE: I guess because the class can still be extended, makes sense then..

File: src/Type/StringType.php [hunk 7]
```diff
@@ -98,19 +98,25 @@ public function accepts(Type $type, bool $strictTypes): TrinaryLogic
...
-			return TrinaryLogic::createFromBoolean(
-				$typeClass->hasNativeMethod('__toString'),
-			);
+		$thatClassNames = $type->getObjectClassNames();
+		if ($thatClassNames === [] || $strictTypes) {
+			return TrinaryLogic::createNo();
 		}
 
-		return TrinaryLogic::createNo();
+		return TrinaryLogic::createNo()->lazyOr(
```
Comment (staabm) [inline src/Type/StringType.php]: sorry if my question is dump,.. I don't foresee the overall change of this PR .. 

I am wondering, whether this change here means, that the new logic accepts a union of types with a `__toString` method, which it did not accept before...?
(in other words, does this change add a feature which needs a new test?)

File: src/Type/ClosureType.php [hunk 8]
```diff
@@ -97,6 +97,11 @@ public function getReferencedClasses(): array
 		return array_merge($classes, $this->returnType->getReferencedClasses());
 	}
 
+	public function getObjectClassNames(): array
```
Comment (staabm) [inline src/Type/ClosureType.php:100]: another maybe dumb question:

~~before this PR the `ClosureType` was not considered a `TypeWithClassName`, additionally it was not handled in `TypeUtil::getDirectClassNames()` etc.~~

but we are now using `Type->getObjectClassNames()` in a few places where we do not handle/did not handle the `ClosureType` type before, but it is now returned.

does this mean we have new features/new edge cases in this area which require new tests?

**scratch this comment.. I did a wrong code-search**

File: src/Type/StringType.php (hunk 7 above)
Comment (herndlm) [reply src/Type/StringType.php]: No, you're right. Things like that were the main reason I wanted to not change too much here and try to get rid of instanceof in follow-ups.
I basically tried to use the new method everywhere instead and this is how I came up with such changes. I'm not sure of all yet, and yes, this one for example should be tested somehow I guess.
I'm still OK with only merging the first commit here. But I can also start looking into adding more tests.

Comment (herndlm) [reply src/Type/StringType.php]: @ondrejmirtes what do you think - I just took a second look at this. some of my adaptions where I tried to handle cases where `$type->getObjectClassNames()` would return more than one entry, are adapted in the same fashion using TrinaryLogin::lazy*(). BUT, AFAIK so far only `UnionType` and `IntersectionType` would return more than one. And e.g. here further up above `CompoundType` is already handled, so this code is basically useless.

I could simplify it again to e.g.
- throw a `ShouldNotHappenException` in case it's more than one and handle it if it's one or zero entries
- always only handle the first entry which would be set or null

not sure if my adaptions like this one make sense. I can't test them right now for sure :/

Comment (herndlm) [reply src/Type/StringType.php]: nevermind, cleaning it up already, I'll mark it as ready to review when I'm done. no need to check now

File: src/Type/ObjectType.php [hunk 9]
```diff
@@ -248,16 +248,22 @@ public function accepts(Type $type, bool $strictTypes): TrinaryLogic
 			return TrinaryLogic::createMaybe();
 		}
 
-		if (!$type instanceof TypeWithClassName) {
+		$thatClassNames = $type->getObjectClassNames();
+		if (count($thatClassNames) > 1) {
+			throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/Type/ObjectType.php:253]: Why it couldn't be more than one? In the world where the only CompoundType is NeverType, this should be supported and implemented. By calling `checkSubclassAcceptability` for each class name and combining the results correctly.

File: src/Type/ObjectType.php [hunk 10]
```diff
@@ -309,20 +315,22 @@ public function isSuperTypeOf(Type $type): TrinaryLogic
 		}
 
 		$thisClassName = $this->className;
-		$thatClassName = $type->getClassName();
+		if (count($thatClassNames) > 1) {
+			throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/Type/ObjectType.php:319]: dtto

File: src/Type/StringType.php [hunk 11]
```diff
@@ -98,19 +100,24 @@ public function accepts(Type $type, bool $strictTypes): TrinaryLogic
...
 		}
 
-		if ($type instanceof TypeWithClassName && !$strictTypes) {
-			$reflectionProvider = ReflectionProviderStaticAccessor::getInstance();
-			if (!$reflectionProvider->hasClass($type->getClassName())) {
-				return TrinaryLogic::createNo();
-			}
+		$thatClassNames = $type->getObjectClassNames();
+		if (count($thatClassNames) > 1) {
+			throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/Type/StringType.php:105]: dtto

File: src/Analyser/MutatingScope.php [hunk 12]
```diff
@@ -2730,6 +2730,11 @@ public function enterClosureCall(Type $thisType, Type $nativeThisType): self
 		$nativeExpressionTypes = $this->nativeExpressionTypes;
 		$nativeExpressionTypes['$this'] = ExpressionTypeHolder::createYes(new Variable('this'), $nativeThisType);
 
+		$thisObjectClassNames = $thisType->getObjectClassNames();
+		if (count($thisObjectClassNames) > 1) {
+			throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php]: Can be more than one. You can typically have intersection type of ThisType and another type, for example:

```php
if ($this instanceof FooInterface) {
    // voila
}
```

You can change MutatingScope internals to correctly represent this.

Comment (ondrejmirtes) [reply src/Analyser/MutatingScope.php]: Should be tested of course :)

File: src/PhpDoc/TypeNodeResolver.php [hunk 13]
```diff
@@ -629,9 +628,14 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
 			return new ErrorType();
 		} elseif ($mainTypeName === 'value-of') {
 			if (count($genericTypes) === 1) { // value-of<ValueType>
-				if ($genericTypes[0] instanceof TypeWithClassName) {
-					if ($this->getReflectionProvider()->hasClass($genericTypes[0]->getClassName())) {
-						$classReflection = $this->getReflectionProvider()->getClass($genericTypes[0]->getClassName());
+				$genericTypeObjectClassNames = $genericTypes[0]->getObjectClassNames();
+				if (count($genericTypeObjectClassNames) > 1) {
+					throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/PhpDoc/TypeNodeResolver.php:633]: I think this can happen by writing something crazy in `value-of`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Awesome, I'm really looking forward to this :)

File: src/Type/ObjectType.php (hunk 9 above)
Comment (herndlm) [reply src/Type/ObjectType.php:253]: I even had something like that already but removed it again 😅 the problem for me was that I couldn't come up with a test (because as you hinted already, CompoundType is already handled before) and I was afraid that this "dead" code might break at some point in the future in weird ways. That's why I thought it might make sense to just not support it for now and break in more expected ways

Comment (ondrejmirtes) [reply src/Type/ObjectType.php:253]: That's fine - write something that works, try it with a unit test and maybe temporarily make UnionType not implement CompoundType to see if your logic works...

When I make most types no longer implement CompoundType, I don't want to hunt for ShouldNotHappenExceptions 😊

File: src/Analyser/MutatingScope.php (hunk 12 above)
Comment (herndlm) [reply src/Analyser/MutatingScope.php]: I pushed a fixup trying to adapt it. I'm not fully sure about it though. I added a test that at least leads now to passing 2 classNames there, which would have thrown an exception before, but the assertion was the same before. In theory something should be smarter now around `canAccessClassMember()` but I couldn't figure out how to test this yet :/

**Comment by staabm** [comment]
Would be great if we could land this one, so we can continue with the refactoring without conflicts

**Comment by ondrejmirtes** [comment]
Yeah, sure, could be out of draft already 😊 This thing is slowing down my mail client, with 50 messages in the thread 😂

**Comment by herndlm** [comment]
Bump 😅

I'll try to undraft it soon™

File: src/Type/ObjectType.php (hunk 9 above)
Comment (herndlm) [reply src/Type/ObjectType.php:253]: this is a good example, I tried to
> By calling checkSubclassAcceptability for each class name and combining the results correctly. 

added test cases and commented out the CompoundType handling. I either had
```php
return TrinaryLogic::createNo()->lazyOr(
	$thatClassNames,
	fn (string $thatClassName) => $this->checkSubclassAcceptability($thatClassName),
);
```

or 

```php
return TrinaryLogic::createYes()->lazyAnd(
	$thatClassNames,
	fn (string $thatClassName) => $this->checkSubclassAcceptability($thatClassName),
);
```

but both are wrong. I don't know anymore if it was a union or an intersection, so I have no clue how to deal with them :/

my added test cases for `accepts()` were
```php
			[
				new ObjectType(DateTimeInterface::class),
				new UnionType([
					new ObjectType(stdClass::class),
					new ObjectType(stdClass::class),
				]),
				TrinaryLogic::createNo(),
			],
			[
				new ObjectType(DateTimeInterface::class),
				new UnionType([
					new ObjectType(stdClass::class),
					new ObjectType(DateTime::class),
				]),
				TrinaryLogic::createMaybe(), // was no with lazyAnd(), was yes with lazyOr()
			],
			[
				new ObjectType(DateTimeInterface::class),
				new UnionType([
					new ObjectType(DateTime::class),
					new ObjectType(DateTime::class),
				]),
				TrinaryLogic::createYes(),
			],
			[
				new ObjectType(DateTimeInterface::class),
				new IntersectionType([
					new ObjectType(stdClass::class),
					new ObjectType(stdClass::class),
				]),
				TrinaryLogic::createNo(),
			],
			[
				new ObjectType(DateTimeInterface::class),
				new IntersectionType([
					new ObjectType(stdClass::class),
					new ObjectType(DateTime::class),
				]),
				TrinaryLogic::createYes(), // was no with lazyAnd()
			],
			[
				new ObjectType(DateTimeInterface::class),
				new IntersectionType([
					new ObjectType(DateTime::class),
					new ObjectType(DateTime::class),
				]),
				TrinaryLogic::createYes(),
			],
```

these are my main problems here.. any ideas?

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by staabm** [comment]
Great job @herndlm 🙏

**Comment by ondrejmirtes** [comment]
An idea - `Type::getStringClassNames()` for these cases:

* ConstantStringType that's also a class string
* GenericClassStringType - a shortcut to `$type->getGenericType()->getObjectClassNames()`

There are some cases of `instanceof GenericClassStringType` that we'd be able to get rid of.
